Procedural animation effects for playlab goal objects

islemaster opened the pull request:
We wanted a way to 'pulse' or 'flash' goal objects to draw attention to them and help them stand out from the rest of the level.  Here, I introduce an image effect system using SVG filters with three starter effects: "pulse", "glow" and "shine."

_Note: For GIFs below I've manually added the effect to a hazard "item" sprite, since our circle placeholder "goal" sprites don't do the effects justice :grin:.  Also, GIF compression doesn't show some of these so well - they are much smoother when rendered._

**Pulse**
A brightness increase to pure white and back.
![filter_pulse_preview](https://cloud.githubusercontent.com/assets/1615761/10707215/083c38b0-79a4-11e5-8a0f-5fd14189b9a2.gif)

**Glow**
A white glow outline that appears and fades.
![filter_glow_preview](https://cloud.githubusercontent.com/assets/1615761/10707218/119a3e2a-79a4-11e5-8c1a-bbcfb346e68b.gif)

**Shine**
A spotlight that sweeps across the object from top-left to lower-right.
![filter_shine_preview](https://cloud.githubusercontent.com/assets/1615761/10707223/1d66a77a-79a4-11e5-9406-509d17b30fe7.gif)

There are two major caveats to this new feature.
1. The animations are currently implemented using SMIL SVG animation, which is unsupported in all versions of IE and deprecated in other browsers.  Before we ship, we need to convert the animation of the filter properties to be driven by Javascript.
2. SVG filters are unsupported in IE9.  We're going to consider it a graceful degradation to just omit these extra visual effects for our IE9 users.  Before we ship, we should add a feature check to the `ImageFilter` base class and turn its `applyTo` method into a no-op for IE9 users.

Otherwise, take these and run with it!  Have fun!

islemaster commented on apps/src/studio/studio.js:
```diff
@@ -374,6 +378,9 @@ var drawMap = function () {
       spriteFinishMarker.setAttribute('clip-path', 'url(#finishClipPath' + i + ')');
+      if (goalFilter) {
+        goalFilter.applyTo(spriteFinishMarker);
+      }
```
So easy to use!

islemaster commented:
cc @cpirich, and @caleybrock and @bcjordan in case this is useful in the tutorials you guys are working on.

breville commented on apps/src/studio/ImageFilter.js:
```diff
@@ -0,0 +1,121 @@
+ * Wrapping the filters this way also provides an easy place to dynamically
+ * manipulate their properties, generating filter animation.
+ *
+ * TODO: SVG filters are not supported in IE9.  This class should do a feature
```
this seems important.  what happens if it's not a no-op in ie9?

breville commented:
This is really cool: nice code, and nice effects too.  I'm concerned around the IE9 incompatibility for all effects and all-IE incompatibility for some of them.  We've seen classrooms with a mixed range of devices, e.g. a bunch of portable Surfaces with IE but also a bunch of desktops running Chrome, and I'm worried about inconsistencies between what different students are seeing, let alone between what a teacher sees/demos and what the students see.  It seems as though we could achieve a number of these effects with a simple spritesheet which would be functional everywhere, and not too burdensome for a content partner to produce.

So my inclination is to maybe hold off on committing this until we know we're going to use it in production.  But it's also a great feature, so if somebody else finds a use for it now, that would be a vote in favour of getting it in production.

islemaster commented:
Sounds good; I'll hold onto this at least until I get the animation working in a way that is IE10+ compatible.

I feel like having a little library of these is so useful, at least for prototyping, that the graceful degradation cost is worth it.  Sure, the effects are simple and a partner could/should easily bake them in, but I don't want to make them do that for every sprite when we can code it once and have it available for any image in playlab.

Besides, having a little framework for the filters system is so powerful - we could build up to things like the 'weathered text' effect here: http://www.smashingmagazine.com/2015/05/why-the-svg-filter-is-awesome/

Maybe the answer is we keep these around for prototyping (like the "scale" feature) but try to get everything baked before we release anything?

islemaster commented:
The goal effects now play on goals when the level loads, and are removed the first time the student presses Run.

@breville @cpirich please take another look.  I've removed all SMIL animation and replaced it with JS-driven animation of the filter attributes, so the animations are now supported in IE10+.  I still need to do the filter feature-detection so we ensure this work is skipped in IE9.

The effects are using their own `setInterval` which is kind of ugly, but I went ahead with it after realizing that I can't hook them up to `Studio.onTick` because I need the animations to run _before_ the student hits the Run button.

breville commented:
seems reasonable to me.   all the usual caveats about compatibility etc. :)

islemaster commented:
Added SVG filter feature-check.
